### PR TITLE
Implement [MSDistribute checkForUpdate]

### PR DESCRIPTION
**Pull request by thyeggman**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Create a checkForUpdate method to force manual check of new release.

## Misc

[AB#75299](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/75299)

### CODE REVIEW

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 1]
```diff
@@ -37,8 +43,12 @@ - (void)testCheckForUpdateAuthenticateWhenItHasNotAuthenticated {
   MSDistribute *distribute = [MSDistribute new];
   id distributeMock = OCMPartialMock(distribute);
   OCMStub([distributeMock canBeUsed]).andReturn(YES);
+  id parserMock = OCMClassMock([MSBasicMachOParser class]);
```
Comment (owlstronaut) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: does this need to stopMocking?

Comment (guperrot) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: we stopMocking for this in other tests indeed

File: AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h [hunk 2]
```diff
@@ -162,6 +162,11 @@ static NSString *const kMSTesterAppUpdateSetupFailedKey = @"MSTesterAppUpdateSet
  */
 @property(nonatomic) MSDistributeFlags distributeFlags;
 
+/**
+ * Flag to indicate whether an update was started with a checkForUpdate call.
+ */
+@property(atomic) bool checkForUpdateFlag;
```
Comment (jaeklim) [inline AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: We can get rid of this property and add a parameter to `startUpdate` for a flag that inforce update flow even though automatic check is disabled. We don't care authentication stuff at the moment so we can simplify this one.

File: AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h [hunk 3]
```diff
@@ -187,17 +187,6 @@ static NSString *const kMSTesterAppUpdateSetupFailedKey = @"MSTesterAppUpdateSet
...
 - (void)disableAutomaticCheckForUpdate;
 
-/**
- * Check the latest release of the application.
- *
- * @param updateToken The update token stored in keychain. This value can be nil if it is public distribution.
- * @param distributionGroupId The distribution group Id in keychain.
- * @param releaseHash The release hash of the current version.
- */
-- (void)checkForUpdateWithUpdateToken:(nullable NSString *)updateToken
```
Comment (jaeklim) [inline AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: This is a rollback as it was added earlier based on different understanding of the spec.

File: AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h [hunk 4]
```diff
@@ -350,9 +339,11 @@ static NSString *const kMSTesterAppUpdateSetupFailedKey = @"MSTesterAppUpdateSet
 - (void)dismissEmbeddedSafari;
 
 /**
- * Start update workflow
+ * Start update workflow.
+ *
+ * @param onStart A flag that indicates whether it is called when Distribute is started or not.
  */
-- (void)startUpdate;
+- (void)startUpdateOnStart:(BOOL)onStart;
```
Comment (jaeklim) [inline AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: We probably want to talk about this method signature even though this is internal use only.

Comment (thyeggman) [reply AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: If the `onStart` parameter means what I think it means, the signature should just be `- (void)startUpdate:(BOOL)onStart`...

But `onStart` is supposed to indicate if the call is happening when the SDK is starting, right?  I think something like `isSdkStarting` would be clearer

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 5]
```diff
@@ -4,111 +4,278 @@
...
+  [MSDistribute disableAutomaticCheckForUpdate];
+
+  // Then
+  XCTAssertFalse(self.sut.automaticCheckForUpdateDisabled);
+}
+
+- (void)testCheckForUpdateOpenBrowserEvenThoughAutomaticUpdateIsDisabled {
+
+  // If
+  id utilityMock = OCMClassMock([MSUtility class]);
```
Comment (thyeggman) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: Utility mock should be shared.  It's used by 4+ tests now and seems to be mostly set up the same way.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 6]
```diff
@@ -4,111 +4,278 @@
...
 
+@property(nonatomic) MSDistribute *sut;
+@property(nonatomic) id settingsMock;
+@property(nonatomic) id bundleMock;
+@property(nonatomic) id keychainUtilMock;
+@property(nonatomic) id parserMock;
+
 @end
 
 @implementation MSDistributeCheckForUpdateTests
```
Comment (thyeggman) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: I don't see any tests that call `startUpdateOnStart:YES`, I would think there should be some.  Or do these tests verify its use some other way without calling it directly?

Comment (jaeklim) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: Last two tests cover this scenario as those are testing from start.

File: AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h (hunk 4 above)
Comment (owlstronaut) [reply AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: `updateShouldBeginOnStart:(BOOL)shouldBegin` could work. I think it is fine as is.

**Comment by owlstronaut** [comment]
What's been done here?
image

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 7]
```diff
@@ -3,112 +3,283 @@
...
+}
+
+- (void)tearDown {
+  [super tearDown];
+  [self.settingsMock stopMocking];
+  [self.keychainUtilMock stopMocking];
+  [self.bundleMock stopMocking];
+  [self.parserMock stopMocking];
+  [self.utilityMock stopMocking];
+  self.sut = nil;
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: Add `resetSharedInstance` to `tearDown`

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 8]
```diff
@@ -3,112 +3,283 @@
...
 
   // If
-  NSString *updateToken = @"UpdateToken";
-  NSString *distributionGroupId = @"DistributionGroupId";
-  NSString *releaseHash = @"ReleaseHash";
-  id distributeMock = OCMPartialMock([MSDistribute new]);
+  NSDictionary<NSString *, id> *plist = @{@"CFBundleShortVersionString" : @"1.0", @"CFBundleVersion" : @"1"};
+  OCMStub([self.bundleMock infoDictionary]).andReturn(plist);
+
+  // Distribute Mock
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]:
```suggestion
  // Distribute mock.
```

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 9]
```diff
@@ -3,112 +3,283 @@
...
+  OCMStub([distributeMock canBeUsed]).andReturn(YES);
+  OCMStub([distributeMock isEnabled]).andReturn(YES);
+  [self.settingsMock removeObjectForKey:kMSUpdateTokenRequestIdKey];
+  OCMStub([distributeMock checkForUpdatesAllowed]).andReturn(YES);
+  XCTestExpectation *expectation = [self expectationWithDescription:@"Start update processed"];
+
+  // Ingestion Mock
+  __block id ingestionMock = OCMClassMock([MSDistributeIngestion class]);
+  OCMStub([ingestionMock checkForPublicUpdateWithQueryStrings:OCMOCK_ANY completionHandler:OCMOCK_ANY]).andDo(^(NSInvocation *invocation) {
+    void (^handler)(NSString *callId, NSHTTPURLResponse *_Nullable response, NSData *_Nullable data, NSError *_Nullable error);
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: Please add typedef to make it more readable

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m [hunk 10]
```diff
@@ -3,112 +3,283 @@
...
-  OCMStub([bundleMock infoDictionary]).andReturn(plist);
-  
-  // Parser mock
-  id parserMock = OCMClassMock([MSBasicMachOParser class]);
-  OCMStub([parserMock machOParserForMainBundle]).andReturn(parserMock);
-  OCMStub([parserMock uuid]).andReturn([[NSUUID alloc] initWithUUIDString:@"CD55E7A9-7AD1-4CA6-B722-3D133F487DA9"]);
-  
+  OCMStub([self.bundleMock infoDictionary]).andReturn(plist);
+
+  // Distribute Mock
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]:
```suggestion
  // Distribute mock.
```

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m (hunk 7 above)
Comment (jaeklim) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: We shouldn't rely on clean up that's done by other tests. We should do it in `setUp` wherever it is required.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m (hunk 9 above)
Comment (jaeklim) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeCheckForUpdateTests.m]: We don't do this in unit tests and just one time use only.

File: AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h (hunk 4 above)
Comment (thyeggman) [reply AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: I don't have a strong preference, if everything else is good we should merge this to unblock wrapper SDKs

Comment (jaeklim) [reply AppCenterDistribute/AppCenterDistribute/MSDistributePrivate.h]: This is still open and we will revisit this soon.

**Pull request merged**